manifests/osfamily/redhat.pp

```diff
@@ -123,11 +113,25 @@
     }
 
     # Given the path to a key, see if it is imported, if not, import it
+    $legacy_gpg_pubkey = "gpg-pubkey-$(echo $(${gpg_cmd} --with-colons ${legacy_gpg_path} 2>&1 | grep ^pub | awk -F ':' '{print \$5}' | cut --characters=9-16 | tr '[:upper:]' '[:lower:]'))"
     $gpg_pubkey = "gpg-pubkey-$(echo $(${gpg_cmd} --with-colons ${gpg_path} 2>&1 | grep ^pub | awk -F ':' '{print \$5}' | cut --characters=9-16 | tr '[:upper:]' '[:lower:]'))"
-    exec {  "import-${keyname}":
+
+    file { "${::env_temp_variable}/gpg_import_check.sh":
+      ensure => file,
+      source => 'puppet:///modules/puppet_agent/gpg_import_check.sh',
+      mode   => '0755',
+    }
+    -> exec { "import-${legacy_keyname}":
+      path      => '/bin:/usr/bin:/sbin:/usr/sbin',
+      command   => "rpm -q ${legacy_gpg_pubkey} && rpm -e --allmatches ${legacy_gpg_pubkey}; rpm --import ${legacy_gpg_path}",
+      unless    => "${::env_temp_variable}/gpg_import_check.sh ${gpg_homedir} ${legacy_gpg_path}",
+      require   => File[$legacy_gpg_path],
+      logoutput => 'on_failure',
+    }
+    -> exec { "import-${keyname}":
       path      => '/bin:/usr/bin:/sbin:/usr/sbin',
-      command   => "rpm --import ${gpg_path}",
-      unless    => "rpm -q ${gpg_pubkey}",
+      command   => "rpm -q ${gpg_pubkey} && rpm -e --allmatches ${gpg_pubkey}; rpm --import ${gpg_path}",
+      unless    => "${::env_temp_variable}/gpg_import_check.sh ${gpg_homedir} ${gpg_path}",
```

Comment:
Could we move all this gpg key mangling into a script `gpg_helper.sh`. If `unless` parameter is still useful, the same script can be called with arguments.